src/libs/Network/SequentialQueue.js

```diff
@@ -18,6 +18,13 @@ let isSequentialQueueRunning = false;
 
 let currentRequest = null;
 
+let resolveIsDonePromise;
+let isDonePromise = new Promise((resolve) => {
```

Comment:
I'm having trouble differentiating `isDonePromise` from `isReadyPromise`. What about these two promises is different? I think they are both initialized and both finishing at the same time.

Reply:
That is true, except the `isReadyPromise` is immediately resolved to when we start the App here, and is also immediately subscribed to every time an object is pushed to the SequentialQueue.

The `isDonePromise` is intended to be used externally with `getIsDonePromise()`.

Reply:
Sorry, also not following. Why does it matter if it is resolved immediately when we start the app? It looks like it gets set again when we flush the queue so can't we use it to tell that we need to wait for the queue to finish?

Reply:
It needs to be resolved immediately so that `flush` is called immediately (via [this](https://github.com/Expensify/App/blob/eab531d9f19aca671bb84b16a44618e4c671899c/src/libs/Network/SequentialQueue.js#L101-L102)).

If we waited `read` requests on the `isReadyPromise` then the `isReadyPromise` would get resolved immediately and `read` requests would fire directly.

Reply:
The `isReadyPromise` stuff was added by @tgolen, perhaps he can do a better job of explaining than I can.

Reply:
@tgolen Why do we resolve the `isReadyPromise` [here](https://github.com/Expensify/App/blob/005240398e3a43328a07ce8791825fd415d4459f/src/libs/Network/SequentialQueue.js#L15-L16) instead of calling `flush` directly instead?

Reply:
@marcaaron Honestly, to add on, I think adding the concept of the `isDonePromise` just makes the SequentialQueue a lot easier to work with.

This entire time I've been looking for a way to get read requests to run after the queue is done. If I saw that the SequentialQueue had a `isDonePromise` combined with an exported `getIsDonePromise` getter, then I'd know immediately that was what I needed.

Reply:
>I think adding the concept of the isDonePromise just makes the SequentialQueue a lot easier to work with.

I know this stuff is **hard**, but I am having trouble agreeing 😄 I think all this promise stuff is getting super confusing and would really love to figure out a way to make this simpler and resist adding complexity (but you are also free to ask for other opinions).

>It needs to be resolved immediately so that flush is called immediately (via [this](https://github.com/Expensify/App/blob/eab531d9f19aca671bb84b16a44618e4c671899c/src/libs/Network/SequentialQueue.js#L101-L102)).

Something about the existing code seems a little off to me (pretty sure I wrote it and Tim just changed some syntax). I am hoping that maybe we can improve how things work if we can understand the motivations behind them a bit better...

as far as I can tell, the `isReadyPromise`'s main purpose is handling the case where a request is "pushed" to the queue and the queue is already running. When this happens, it will get processed _after_ the current run of the queue is done. So it's more like an "is queue done" promise. We resolve it when the queue is finished. Which then allows the next run of the queue to happen (requests pushed while the queue is running will wait for the current batch to finish).

The comment about the queue starting when the page loads is throwing me off a little (I don't remember writing that). I _think_ the reason we resolve it immediately is because if a `push()` were to get called before a `flush()` then we will be waiting for a promise that never resolves. When `flush()` gets called here it will replace the `isReadyPromise` entirely and leave the call to `push()` waiting for a promise that will never resolve. Hard to say which would happen first or if we did that to actively or preemptively solve an issue, but in either case I'd imagine we'd have these two possibilities:

**Scenario 1:**

flush() -> Queue now running and resets the `isReadyPromise` value
push() -> now waiting for the current run to finish

**Scenario 2:**

push() -> Queue not running but promise is resolved so we can call flush() right away

By resolving the promise immediately we will basically let a potential first call to `push()` determine whether to start the queue or not. Otherwise, it gets started here:

https://github.com/Expensify/App/blob/847e0af9f05756f49756cc374d0aa5aed02118a0/src/libs/Network/index.js#L8-L10

So, stepping back again, it seems like the requirements we have are:

1. The sequential queue must run when the app first inits and before any read requests run
2. Any read requests must wait until the sequential queue is 100% done processing before they can proceed

Definitely feels like we can come up with some way to have that happen more reliably. What's not super clear to me is the relationship between the read requests that we are wanting to defer and the running of the sequential queue requests. It kind of still sounds to me like we are just calling them before the sequential queue runs - which is just how I would expect things to behave.

Reply:
I agree with Marc on most of what he said here. 

I can maybe add a little more context as to why the isReadyPromise is immediately resolved (though, I didn't write it, it's just how I understand it). Marc is correct with his "Scenario 2" illustration. The first time that `push()` is called, the request won't happen until `isReadyPromise.then(flush);`. This creates a chicken-and-egg problem. The request from `push()` won't run until the promise is resolved and flush is triggered, and the promise is only resolved when `flush()` is done. Therefore, immediately resolving the promise when the app starts allows the first `push()` to be processed because the promise is already resolved, even though `flush()` hasn't happened yet. In effect, we're forcing the chicken to lay the egg.

I think we're getting close, but I agree that having two promises that do pretty much the same thing, is more confusing and complex than what exists today, so it feels like we're taking two steps back and only one step forward.

If we can combine them into a single promise, and expose it externally, then that would be great.

Reply:
@tgolen @marcaaron thank you guys for the explanation, after hearing your explanations of the intended purpose of the `isReadyPromise` I can fully agree that the `isDonePromise` is extraneous. Perhaps the name `isReadyPromise` can be updated to more accurately reflect what we're exposing it and using it for.

> If we can combine them into a single promise, and expose it externally, then that would be great.

That sounds like a plan, I'll continue experimenting with using `isReadyPromise`

Reply:
I'm thinking if we can use the `isReadyPromise`, but expose it in a function called something like `runAfterDone`? Are there any other situations in which we'd want to wait on the SequentialQueue to finish running?

Reply:
> the read requests that we are wanting to defer and the running of the sequential queue requests. It kind of still sounds to me like we are just calling them before the sequential queue runs - which is just how I would expect things to behave.

@marcaaron can you clarify what you're confused about? Is it the current relationship of between read requests and the SeqQueue on `main` or this branch?

Reply:
Mostly confused about which requests we want to defer specifically and the general order of operations. It might help to write that all down so we can summarize which "read" requests happen before which "write" requests and what triggers them to happen. I maybe missed some earlier parts of this conversation, but I'm holding an incomplete picture of the problem we are trying to address.

Reply:
I think in most cases we want read requests to be held on the SequentialQueue, because that's what's best for most write requests. You'll want to let the write requests finish first so their optimistic data doesn't get overwritten by lightweight, fast-responding read requests using the same Onyx key. For the same Onyx key, write requests almost always take longer to respond than read requests. This is what causes the flickering effect.

If there are any specific read requests that need to run directly, perhaps we can expose different functions API functions for the different cases? I guess I'm not entirely sure what these would look like.

Reply:
Maybe it would be more productive to have a Slack discussion or meeting about this topic. I'm still unsure what _specific requests_ you are looking at while testing this PR/issue. And again, it sounds like they are getting triggered **before** the write queue starts - but hard to tell that without debugging this myself and doing the work. 

So, I’m just kinda confused and looking for a specific practical explanation of what we are trying to fix not the high level explanation (I think everyone is on the same page there and agrees that reads should wait for writes to finish).

Reply:
@marcaaron The practical explanation of what we're trying to fix (hopefully I'm understanding you correctly) is that when you make edits to workspaces offline and then come back online, the values you've modified will temporarily revert back to the old values, and then will go back to the new value after a few seconds.

This is due to how the optimistic data set by the write request to update the workspace offline gets temporarily overwritten by the response from read requests that are fired immediately when we come back online. Specifically, I've been testing on the `WorkspaceReimburseView` (Workspace > Reimburse expenses). I go offline, modify either the custom unit or custom unit rate, and come back online. I want to ensure that `UpdateWorkspaceCustomUnit` (write) responds before `OpenWorkspaceReimburseView` (read) runs.

Reply:
Got it, thanks! 

At the issue level, it feels like we need some way to ensure that `UpdateWorkspaceCustomUnit` always happens before `OpenWorkspaceReimburseView` and at a higher level we can solve that by making all "reads" finish before any "writes" if they are queued in close proximity.

Dug a little deeper and requests to update the rate are going into the network queue so they should be fired when we come back from online here:

https://github.com/Expensify/App/blob/ec5d51c577eaeaa30d48bbd7c250e977a60ac33b/src/libs/Network/SequentialQueue.js#L84-L84

But the reconnect logic on this page will fire the "read" request first since nothing stops it from happening:

https://github.com/Expensify/App/blob/ec5d51c577eaeaa30d48bbd7c250e977a60ac33b/src/pages/workspace/reimburse/WorkspaceReimburseView.js#L109-L114

They should both be happening at roughly the same time (or maybe the exact same time) so we can reliably hold any "reads" on the "writes" and avoid all these conflicts in the future (hopefully).